### DISCUSSION TITLE
Nginx Proxy Manager

### DISCUSSION THREAD

**Discussion by MayhemBill** [mastodon/mastodon#18799] [Q&A]
I have been using regular nginx for a while as a reverse proxy and recently swapped to nginx proxy manager, but given Mastodons fairly complicated (at least compared to everything else I use), it seems a big more difficult than just point it at the port 3000 and it just working, has anyone else been able to get it working and mind sharing their configs?

**Comment by MayhemBill** [comment]
So after a bit of tinkering, I have boiled the first issue down to Mastodon-web running on 127.0.0.1, and the docker container for NPM not being able to access it.

**Comment by valon0388** [comment]
Currently trying to set up the same thing. I've modified the Mastodon instances to be open to the reverse proxy, but still can't get the site to resolve. Did you have any luck here?

**Reply by redimongo** [reply]
How did you do that?

**Reply by valon0388** [reply]
Unfortunately, I don't remember at this point. I need to revisit this, as I still want to host my own instance; but I gave up on that attempt due to me not having the homelab set up correctly to serve the content at this time. My next attempt is going to be on a Digital Ocean instance I have set up. Maybe it will go better.

**Comment by fampla** [comment]
Sorry for the stupid question, but you have your domain only directed to the _web_1 and in advanced nothing else entered?

Oh, of course, all external to the same network.
I ask because I do not really get on right now -.-"

**Reply by fampla** [reply]
Thanks a lot. Possible that I forgot the /api/ /api/v1/streaming.
Thanks a lot, now it is working 😀

**Reply by PhoenixSheppy** [reply]
Thank you so much! This is super useful!

**Reply by rogawa** [reply]
Thank you @pedroCX486. Works like a charm!

**Comment by EasternPA** [comment]
@pedroCX486 thank you for this. I'm about 8 hours into trying to stand this up on an arm cloud instance behind NPM and cloudflare proxy. I already run a bunch of services through CF and NPM w/LE, so it shouldn't be this hard. With your help with the /api custom location, I finally got a web page to come up. The only page that comes up would be my profile page but it is 404 since I can't login and set it up! Is there a home page for the server? Any tip for a URL path to log in?

**Reply by EasternPA** [reply]
Thanks for the reply. I tried a couple of different methods based on docker-compose. I am running a normal instance with signups disabled. I ended up getting beyond my stuck point, but in a really weird way. I never saw the server setup after creating my first userid. Also, my web page always redirected to my admin userid profile page, but it was 404.

I was finally able to get beyond it by creating a second (non-admin) userid and pulling up the profile page. Only then was I prompted with a normal page and the ability to login. I then reset the password on my admin account via the CLI, logged out of the non-admin account in the browser, and logged back in as the admin user. My instance's about page is no longer 404, but my admin account is marked as an admin user but NOT marked as the admin of the instance! Now I just need to figure out how to set my admin id as the instance admin. I'm sure I should see a Settings page when logged in with my admin account, but I don't.

How do I get to the instance admin page? How do I show my admin user as the admin of the instance? I have the admin tag on the id but the instance admin page says no admin assigned yet.

Edit: Found it! Under Preferences on the right, Administration on the bottom left. It almost looks like a real instance now!

**Reply by Cyb3r14n-Security** [reply]
how did you get this working? I've tried the same setup as pedro suggested with the custom location and the advanced tab but its giving me a 'internal error' message and in the logs it says some challenges have failed. the logs for all the mastodon containers look like they're working fine I'm also trying to get let's encrypt to work idk if theres any additional settings in the ssl tab that i need to enable

**Reply by EasternPA** [reply]
Have you ever gotten the web UI for NPM to work with Let's Encrypt? Is there someplace else we can chat without tying up the post? Are you on a different Mastodon instance while you work to stand up your own?

**Reply by Cyb3r14n-Security** [reply]
i use matrix @tazy:matrix.tazy.xyz, my NPM instance has a few different services behind it that work matrix being one of them so i has worked before for other services im not sure if i've properly configured it right in the custom locations is that the only thing thats supposed to be in it the /api/? Also what is the mastodon_sidekiq_1 container? do i need NPM config for that?

**Comment by EasternPA** [comment]
Hey all, I tried to document the entire process of buying a domain, getting a free VPS and installing mastodon with NPM using the web UI. Check it out and leave some feedback.

https://easternpa.github.io/mastodon-install/

Edit: Now that it's been several days since I wrote this, there are some things I would change. As best I can tell, you only need to git clone and copy the docker-compose.yml to another folder, like /opt/mastodon. The install script builds the .env.production for you and my instructions cover copying the content into the file. If you're going to perform the install in /opt/mastodon, then some paths in the commands I wrote will change. I'll get on that.

**Reply by SiskoUrso** [reply]
Thanks for the write up, this helped me finally get my stack up and working correctly.

**Reply by EasternPA** [reply]
Thanks for the thanks. I appreciate it. I'm sure there are better instructions out there. I had a lot of hope for Wildebeest, but it didn't take off as quickly as I thought it would. Plus, it still doesn't allow account transfers (afaik). That isn't good. I also want the Cloudflare team to add more apps to their ActivityPub support, like Bookwyrm and Pixelfed.

My biggest issue with my setup is that Oracle gives you so little space for free. I have cron jobs to delete almost everything every day or two. Nowadays, I read hashtag searches in an RSS reader and copy/paste post URLs into my home instance if I want to interact with them. I recently bought a small form factor PC with 32GB RAM that I intend to use for hosting all my self-hosted apps. The free Oracle stuff was good for getting off the ground, but you quickly run out of space.

**Comment by ghost** [comment]
How does the nginx config on your Mastodon servers look like?

**Reply by EasternPA** [reply]
I don't have one, actually. I prefer to use the NPM GUI, so I just add a Proxy Host. I don't have any custom locations or custom settings in the GUI, either. I just point the Proxy Host for my FQDN to my Mastodon instance and it works fine. I also finally sat down to get Cloudflare Tunnels working, and I finally understand why other people would pop in on discussions like these and say, "Just use a tunnel." It's way easier and makes hosting from home so easy even if your ISP doesn't allow you to open 80 or 443.

**Reply by ghost** [reply]
So I shouldn't even have installed nginx on the mastodon server?

**Reply by EasternPA** [reply]
I have a docker compose yml that I copy and paste first whenever I start a new stack. I don't install it in the Mastodon container
